File: python/FRST.py
import numpy as np
import scipy
from scipy import special


def nufft_check_dft(gamma, Nd):
    M = np.size(gamma)
    Nd = 64

    kk = gamma/(2*np.pi) * Nd
    # kk = gamma/(2*np.pi) * np.tile(Nd[:], (M, 1)] simile al testo ma qui non ha senso dato che Nd e' un intero

    tol = 1e-6
    tmp = np.abs(np.around(kk) - kk)
    if np.all(tmp[:] < tol) and np.any(gamma):
        print('DFT samples has suboptimal accuracy')

def kaiser_bessel(u,J):
    #if J is None:
    #    J = 6
    d = 1
    kb_m = 0
    alpha = 2.34 * J
    z = np.sqrt(np.power(2*np.pi * (J/2)*u,2)-np.power(alpha,2)+0j)
    nu = d/2 + kb_m
    y = np.power(2*np.pi, d/2) * np.power(J/2, d) * np.power(alpha, kb_m) / scipy.special.iv(kb_m, alpha) * \
        scipy.special.jv(nu, z) / np.power(z, nu)
    return np.real(y)

def nufft_alpha_kb_fit(N, J, K):
    beta = 1
    argNmid = (N-1)/2

    if N > 40:
        argL = 13 # empirically found to be reasonable
    else:
        argL = np.ceil(N/3) # a kludge to avoid "rank deficient" complaints

    nlist = np.arange(0,N) - argNmid

    # kaiser-bessel with previously numerically-optimized shape
    sn_kaiser = 1/kaiser_bessel(nlist/K, J)

    # use regression to match NUFFT with BEST kaiser scaling's
    gam = 2 * np.pi /K
    X = np.cos(beta*gam*np.outer(nlist,np.arange(0, argL+1)))
    coef,residuals, rank, s = np.linalg.lstsq(X, sn_kaiser, rcond=None)

    if np.any(np.isnan(coef)): # if any NaN then big problem!
        coef = np.linalg.pinv(X)@sn_kaiser
        if np.any(np.isnan(coef)):
            raise ValueError('bug: NaN coefficients')

    alphas = np.append(np.real(coef[0]), np.real(coef[1:])/2)
    return alphas, beta

def nufft_scale(Nd, Kd, alpha, beta ,Nmid):
    if not np.isreal(alpha[0]):
        raise ValueError('alpha[0] needs to be real')
    L = np.size(alpha)-1 #check

    if L > 0:
        sn = np.zeros((Nd, 1))
        n = np.arange(0, Nd)
        i_gam_n_n0 = 1j * (2*np.pi/Kd) * (n-Nmid) * beta
        i_gam_n_n0 = i_gam_n_n0.reshape((Nd, 1))  # Reshape 1D to 2D array (specific to numpy)

        for l1 in range(-L, L+1):
            alf = alpha[abs(l1)]
            if l1 < 0:
                alf = np.conj(alf)
            sn = sn + (alf * np.exp(i_gam_n_n0 * l1))
    else:
        sn = alpha * np.ones((Nd, 1))
    return sn


def nufft_T(N, J, K, tol, alpha, beta):

    if N > K:
        raise ValueError('N > K error')
    try:
        alpha
    except NameError:
        j2, j1 = np.meshgrid(np.arange(1, J + 1), np.arange(1, J + 1))
        cssc = np.sinc((j2 - j1) / (K / N))

    if alpha is None:
        j2, j1 = np.meshgrid(np.arange(1, J + 1), np.arange(1, J + 1))
        cssc = np.sinc((j2 - j1) / (K / N))

    # Fourier-series based scaling factors
    else:
        if not np.isreal(alpha[0]):
            raise ValueError('alpha[0] must be real')
        L = np.size(alpha) - 1
        cssc = np.zeros((J, J))
        j2, j1 = np.meshgrid(np.arange(1, J + 1), np.arange(1, J + 1))

        for l1 in range(-L, L + 1):
            for l2 in range(-L, L + 1):
                alf1 = alpha[abs(l1)]
                if l1 < 0:
                    alf1 = np.conj(alf1)
                alf2 = alpha[abs(l2)]
                if l2 < 0:
                    alf2 = np.conj(alf2)
                tmp = j2 - j1 + beta * (l1 - l2)
                tmp = np.sinc(tmp / (K / N))
                cssc = cssc + alf1 * np.conj(alf2) * tmp

    # Inverse (or pseudo inverse)

    u, s, vh = np.linalg.svd(cssc)
    smin = np.min(s)

    if smin < tol:
        print('Warning, poor conditioning ' + str(smin) + ' -> pinverse')
        T = np.linalg.pinv(cssc, rcond=tol / 10)
    else:
        T = np.linalg.pinv(cssc)

    return T

def nufft_r(om, N, J, K, alpha, beta):

    try:
        alpha
    except NameError and alpha is None:
        alpha = 1  # default Fourier series coefficients of scaling factors

    if alpha is None:
        alpha = 1  # default Fourier series coefficients of scaling factors

    try:
        beta
    except NameError and alpha is None:
        beta = 0.5  # default is Liu version for now

    if beta is None:
        beta = 0.5  # default is Liu version for now

    M = np.size(om)
    gam = 2 * np.pi / K
    dk = om / gam - np.floor(om / gam - J / 2)
    arg = np.add.outer(-np.arange(1, J + 1), dk)

    L = np.size(alpha) - 1  # check

    if L > 0:
        rr = np.zeros((J, M))

        for l1 in range(-L, L + 1):
            alf = alpha[abs(l1)]
            if l1 < 0:
                alf = np.conj(alf)
            r1 = np.sinc((arg + l1 * beta) / (K / N))
            rr = rr + alf * r1
    else:
        rr = np.sinc(arg / (K / N))

    return rr, arg


def nufft_interpMatrixKBminMax(gamma, Nd, Jd, Kd):

    M = np.size(gamma)

    nufft_check_dft(gamma, Nd)
    alpha = None
    alpha, beta = nufft_alpha_kb_fit(Nd, Jd, Kd)

    # Find NUFFT scaling factor

    Nmid = (Nd - 1) / 2
    tmp = nufft_scale(Nd, Kd, alpha, beta, Nmid)
    sn = scipy.sparse.coo_matrix((tmp.reshape(Nd), (np.arange(0, Nd), np.arange(0, Nd))), shape=(Nd, Nd))

    # NUFFT interpolation matrix
    tol = 0
    T = nufft_T(Nd, Jd, Kd, tol, alpha, beta)
    r, arg = nufft_r(gamma[:], Nd, Jd, Kd, alpha, beta);
    c = T @ r
    # T = None
    # r = None
    gam = 2 * np.pi / Kd
    phase_scale = 1j * gam * (Nd - 1) / 2

    phase = np.exp(phase_scale * arg)
    uu = np.conj(phase * c)

    # indices into oversampled FFT components
    koff = np.floor(gamma[:] / gam - Jd / 2)
    kk = np.mod(np.add.outer(np.arange(1, Jd + 1), koff), Kd)

    mm = np.arange(0, M)
    mm = np.tile(mm, (np.prod(Jd), 1))

    sp = scipy.sparse.coo_matrix((uu.ravel(), (mm.ravel(), kk.ravel())), shape=(M, np.prod(Kd)))

    return Nd, Kd, M, sn, sp



def nufft_interpMatrixNN(gamma,Nd,Kd):
    M  = np.size(gamma)

    # NUFFT scaling factor
    sn = scipy.sparse.identity(Nd)

    # NUFFT interpolation matrix

    positiveInput = gamma > 0
    gamma = np.mod(gamma, 2*np.pi)
    gamma[(gamma == 0) & positiveInput] = 2*np.pi
    DFT_w = np.linspace(0, 2*np.pi/Kd * (Kd-1), Kd)
    tmp = np.add.outer(gamma, -DFT_w)
    idx_min = np.argmin(np.abs(tmp), axis=1)

    sp = scipy.sparse.coo_matrix((np.ones(M), (np.arange(0, M), idx_min)), shape=(M, Kd))
    return Nd, Kd, M, sn, sp


def nufft(x, Kd,sp):
    # NUFFT
    Xk = np.fft.fft(x, Kd, axis=0)
    X = sp @ Xk

    return X


# FRST
def frst(p, f, c, d, L, mbar, W, qbar, sigma, N, B):

    # Parameters
    z = np.linspace(0, d*(L-1), num=L).T                                  # [L,1] microphone positions
    m = np.linspace(-(W-1)/2*mbar, ((W-1)*mbar-((W-1)/2*mbar)), num=W).T  # [W,1] m axis
    q = np.linspace(0, z[-1], num=np.around(z[-1]/qbar) + 1).T            # [j,1] q axis
    I = np.size(q)                                                        # Number of frames

    # Signal windowing
    arg_psi = np.add.outer(z,-q)
    psi = np.exp(-np.pi*np.power(arg_psi, 2) / np.power(sigma, 2))        # Gaussian window with std sigma

    # Set of uniformly spaced frequency locations
    gamma = (2*np.pi*f/c)*(m/np.sqrt(1+np.power(m, 2)))*d

    if B == 1:
        Nd, Kd, M, sn, sp = nufft_interpMatrixNN(gamma, L, N)
    else:
        Nd, Kd, M, sn, sp = nufft_interpMatrixKBminMax(gamma, L, B, N)

    pbar = np.tile(d*sn@p, (1, I)) * psi


    Z = nufft(pbar, Kd, sp).T

    return Z, m, q, z

File: matlab/FRST.m
function [Z,m,q,z] = FRST(p,f,c,d,L,mbar,W,qbar,sigma,N,B)
%|
%| in
%|	p		[L,1] microphone signal in the frequency domain
%|	f		frequency in Hz
%|	c		speed of sound
%|	d		distance between adjacent microphones
%|	L       Number of microphones
%|	mbar    m axis sampling interval
%|  W       length of m axis
%|	qbar	q axis sampling interval
%|	sigma   gaussian window standar deviation
%|	N       oversampling factor (length of FFT)
%|	B       Number of nearest neighbors for the interpolation
%|          B = 1 (nearest neighbor interpolation),
%|          B > 1 Keiser-Bessel minmax interpolation  see [1]
%|
%| out
%|	Z       [W,I] FRST output
%|	m       [W,1] m axis
%|	q       [I,1] q axis
%|  z       [L,1] microphone positions
%|
%| [1] J. A. Fessler and B. P. Sutton,Nonuniform fast fourier transforms  using  min-max  interpolation
%|

%% Parameters

z = (0:d:d*(L-1))';                         % [L,1] microphone positions
m = ((0:mbar:(W-1)*mbar)-((W-1)/2*mbar))';  % [W,1] m axis
q = (0:qbar:z(end))';                       % [I,1] q axis
I = length(q);                              % number of frames

%% Signal Windowing

arg_psi = outer_sum(z,-q);
psi = exp(-pi*(arg_psi).^2 / sigma^2);      % Gaussian window with std sigma

%% FRST

% Set of nonuniformy spaced frequency locations
gamma = (2*pi*f/c)*((m)./(sqrt(1+m.^2)))*d;

if B == 1
    % Nearest Neighbor interpolation matrix
    st = nufft_interpMatrixNN(gamma,L,N);
else
    % Keiser-Bessel minmax interpolation matrix see [1]
    st = nufft_interpMatrixKBminMax(gamma, L, B, N);
end

% Input signal scaling and windowing
pbar = repmat(d*st.sn*p,1,I).*psi;

% FRST output computed with the Nonuniform Fast Fourier Transform
Z = nufft(pbar,st).';

end


function y = kaiser_bessel(u,J)

if ~exist('J','var'), J = 6; end

d = 1;
kb_m = 0;
alpha = 2.34 * J;

z = sqrt( (2*pi*(J/2)*u).^2 - alpha^2 );
nu = d/2 + kb_m;
y = (2*pi)^(d/2) .* (J/2)^d .* alpha^kb_m ./ besseli(kb_m, alpha) ...
    .* besselj(nu, z) ./ z.^nu;
y = real(y);
end

function X = nufft(x, st)
%function X = nufft(x, st)
%|
%| Compute d-dimensional NUFFT of signal/image x
%|
%| in
%|	x	[N1 N2 ... Nd (L)]	L input image(s) of size
%|						N1 x N2 x ... x Nd
%|	st	structure		precomputed by nufft_init()
%| out
%|	X	[M (L)]			output spectra
%|
%| Copyright 2003-5-30, Jeff Fessler, University of Michigan

%y = st.sn*x;
Xk = fft(x, st.Kd);
X = st.p * Xk;
end

function st = nufft_interpMatrixKBminMax(gamma, Nd, Jd, Kd)

%% Input Check

st.Nd = Nd;
st.Kd = Kd;
M = size(gamma,1);
st.M = M;

nufft_check_dft(gamma, Nd)

%% Find Keiser-Bessel alpha beta parameters

[alpha, beta] = nufft_alpha_kb_fit(Nd, Jd, Kd);

%% Find NUFFT scaling factor

Nmid = (Nd-1)/2;
tmp = nufft_scale(Nd, Kd, alpha, beta ,Nmid);
st.sn = sparse(1:Nd,1:Nd,tmp(:),Nd,Nd);

%% NUFFT interpolation matrix

tol = 0;
T = nufft_T(Nd, Jd, Kd, tol, alpha, beta);
[r, arg] = nufft_r(gamma(:), Nd, Jd, Kd, alpha, beta);
c = T * r;	clear T r
gam = 2*pi/Kd;
phase_scale = 1i * gam * (Nd-1)/2;

phase = exp(phase_scale * arg);
uu = phase .* c;
uu = conj(uu);

% indices into oversampled FFT components
koff = floor(gamma(:) / gam - Jd/2);
kk = mod(outer_sum([1:Jd]', koff'), Kd) + 1;

mm = [1:M];
mm = mm(ones(prod(Jd),1),:); % [*Jd M]
% make sparse matrix
st.p = sparse(mm(:), double(kk(:)), double(uu(:)), M, prod(Kd));
end

function nufft_check_dft(gamma, Nd)
kk = gamma / (2*pi) .* repmat(Nd(:)', [size(gamma,1) 1]);
tol = 1e-6;
tmp = abs(round(kk) - kk);
if all(tmp(:) < tol) && any(gamma(:))
    warning('DFT samples has suboptimal accuracy')
end
end

function [alphas, beta] = nufft_alpha_kb_fit(N, J, K, varargin)

beta = 1;
arg.Nmid = (N-1)/2;

if N > 40
    arg.L = 13;		% empirically found to be reasonable
else
    arg.L = ceil(N/3);	% a kludge to avoid "rank deficient" complaints
end

nlist = [0:(N-1)]' - arg.Nmid;

% kaiser-bessel with previously numerically-optimized shape
sn_kaiser = 1 ./kaiser_bessel(nlist/K, J);

% use regression to match NUFFT with BEST kaiser scaling's
gam = 2*pi/K;
X = cos(beta * gam * nlist * [0:arg.L]); % [N L]
% coef = regress(sn_kaiser, X)';
coef = (X \ sn_kaiser)'; % this line sometimes generates precision warnings
if any(isnan(coef(:))) % if any NaN then big problem!
    coef = (pinv(X) * sn_kaiser)';
    if any(isnan(coef(:)))
        error 'bug: NaN coefficients';
    end
end
alphas = [real(coef(1)) coef(2:end)/2];
end

function sn = nufft_scale(N, K, alpha, beta, Nmid)

if ~isreal(alpha(1)), error 'need real alpha_0', end
L = length(alpha) - 1;

if L > 0
    sn = zeros(N,1);
    n = [0:(N-1)]';
    i_gam_n_n0 = 1i * (2*pi/K) * (n - Nmid) * beta;
    
    for l1=-L:L
        alf = alpha(abs(l1)+1);
        if l1 < 0, alf = conj(alf); end
        sn = sn + alf * exp(i_gam_n_n0 * l1);
    end
    
else
    sn = alpha * ones(N,1);
end
end

function st = nufft_interpMatrixNN(om,Nd,Kd)

%% Input

st.Nd = Nd;
st.Kd = Kd;
M = size(om,1);
st.M = M;

%% NUFFT scaling factor
st.sn = sparse(1:Nd,1:Nd,ones(Nd,1),Nd,Nd);
%% NUFFT interpolation matrix

om = wrapTo2Pi(om(:));
DFT_w = 0:2*pi/Kd:2*pi/Kd*(Kd-1);
tmp = outer_sum(om,-DFT_w);
[~,idx_min] = min(abs(tmp),[],2);

uu = ones(M,1);
mm = [1:M];
st.p = sparse(mm(:),idx_min,uu,M,Kd);
end

function lambda = wrapTo2Pi(lambda)
positiveInput = (lambda > 0);
lambda = mod(lambda, 2*pi);
lambda((lambda == 0) & positiveInput) = 2*pi;
end

function [rr, arg] = nufft_r(om, N, J, K, alpha, beta)
%function [rr, arg] = nufft_r(om, N, J, K, alpha, beta, use_true_diric)
%|
%| make NUFFT "r" vector
%|
%| in
%|	om	[M 1]	digital frequency omega in radians
%|	N		signal length
%|	J		# of neighbors used per frequency location
%|	K		FFT size (should be > N)
%|	alpha	[0:L]	Fourier series coefficients of scaling factors
%|	beta		scale gamma=2pi/K by this in Fourier series
%|				typically is K/N (me) or 0.5 (Liu)
%| out
%|	rr	[J M]	r vector for each frequency
%|	arg	[J M]	dirac argument for t=0
%|
%| Copyright 2001-12-13, Jeff Fessler, University of Michigan

if ~exist('alpha','var') || isempty(alpha)
    alpha = [1]; % default Fourier series coefficients of scaling factors
end
if ~exist('beta','var') || isempty(beta)
    beta = 0.5; % default is Liu version for now
end

M = length(om);
gam = 2*pi/K;
dk = om / gam - floor(om(:) / gam - J/2);
arg = outer_sum(-[1:J]', dk');			% [J M] diric arg for t=0

L = length(alpha) - 1;
if ~isreal(alpha(1)), error 'need real alpha_0', end

if L > 0
    rr = zeros(J,M);
    for l1 = -L:L
        alf = alpha(abs(l1)+1);
        if l1<0, alf = conj(alf); end
        r1 = nufft_sinc((arg + l1 * beta) / (K/N));
        rr = rr + alf * r1;			% [J M]
    end
else
    rr = nufft_sinc(arg / (K/N));
end
end

function y = nufft_sinc(x)
%function y = nufft_sinc(x)
%|
%| my version of "sinc" function, because matlab's sinc() is in a toolbox
%|

iz = find(x == 0); % indices of zero arguments
x(iz) = 1;
y = sin(pi*x) ./ (pi*x);
y(iz) = 1;
end

function T = nufft_T(N, J, K, tol, alpha, beta)
%function T = nufft_T(N, J, K, tol, alpha, beta, use_true_diric)
%|
%| Precompute the matrix T = [C' S S' C]\inv used in NUFFT.
%| This can be precomputed, being independent of frequency location.
%|
%| in
%|	N		# signal length
%|	J		# of neighbors
%|	K		# FFT length
%|	tol		tolerance for smallest eigenvalue
%|	alpha	[L+1]	Fourier coefficient vector for scaling
%|	beta		scale gamma=2*pi/K by this for Fourier series
%|
%| out
%|	T	[J J]	precomputed matrix
%|
%| Copyright 2000-1-9, Jeff Fessler, University of Michigan

if ~exist('tol','var') || isempty(tol)
    tol = 1e-7;
end
if ~exist('beta','var') || isempty(beta)
    beta = 1/2;
end

if N > K, fail 'N > K', end

% default with unity scaling factors
if ~exist('alpha','var') || isempty(alpha)
    
    % compute C'SS'C = C'C
    [j1,j2] = ndgrid(1:J, 1:J);
    cssc = nufft_sinc((j2 - j1) / (K/N));
    
    % Fourier-series based scaling factors
else
    if ~isreal(alpha(1)), fail 'need real alpha_0', end
    L = length(alpha) - 1; % L
    cssc = zeros(J,J);
    [j1,j2] = ndgrid(1:J, 1:J);
    for l1 = -L:L
        for l2 = -L:L
            alf1 = alpha(abs(l1)+1);
            if l1 < 0, alf1 = conj(alf1); end
            alf2 = alpha(abs(l2)+1);
            if l2 < 0, alf2 = conj(alf2); end
            
            tmp = j2 - j1 + beta * (l1 - l2);
            tmp = nufft_sinc((tmp) / (K/N));
            cssc = cssc + alf1 * conj(alf2) * tmp;
        end
    end
end


% Inverse, or, pseudo-inverse

%smin = svds(cssc,1,0);
smin = min(svd(cssc));
if smin < tol % smallest singular value
    warning('Poor conditioning %g => pinverse', smin)
    T = pinv(cssc, tol/10);
else
    T = inv(cssc);
end
end

function ss = outer_sum(xx,yy)
%|function ss = outer_sum(xx,yy)
%|
%| compute an "outer sum" x + y'
%| that is analogous to the "outer product" x * y'
%|
%| in
%|	xx	[nx 1]
%|	yy	[1 ny]
%|		more generally: xx [(dim)] + yy [L,1] -> xx [(dim) LL]
%| out
%|	ss [nx ny]	ss(i,j) = xx(i) + yy(j)
%|
%| Copyright 2001, Jeff Fessler, University of Michigan

% for 1D vectors, allow rows or cols for backward compatibility
if ndims(xx) == 2 && min(size(xx)) == 1 && ndims(yy) == 2 && min(size(yy)) == 1
    nx = length(xx);
    ny = length(yy);
    xx = repmat(xx(:), [1 ny]);
    yy = repmat(yy(:)', [nx 1]);
    ss = xx + yy;
end
end